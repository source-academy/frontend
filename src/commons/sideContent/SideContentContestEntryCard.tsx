import { Card, Classes, Elevation, NumericInput, Pre } from '@blueprintjs/core';
import classNames from 'classnames';
import React from 'react';

import { ContestEntry } from '../assessment/AssessmentTypes';

type SideContentConstestEntryCardProps = DispatchProps & StateProps;

type DispatchProps = {
  handleContestEntryClick: (submissionId: number, answer: string) => void;
};

type StateProps = {
  handleVotingSubmissionChange: (entryId: number, rank: number) => void;
  canSave: boolean;
  isValid: boolean;
  contestEntry: ContestEntry;
  entryNumber: number;
  minScore: number;
  maxScore: number;
};

/**
 * Pure Contest Entry Card to to provide numeric input box for contest entry voting.
 * @param props functions for handling input and contest entry details tied to contest entry card.
 * @returns card which provides numeric input to vote for contest entry.
 */
const SideContentContestEntryCard: React.FunctionComponent<
  SideContentConstestEntryCardProps
> = props => {
  const {
    canSave,
    isValid,
    handleContestEntryClick,
    handleVotingSubmissionChange,
    contestEntry,
    entryNumber,
    minScore,
    maxScore
  } = props;

  return (
    <div className={classNames('ContestEntryCard', { wrong: !isValid })}>
      <Card
        className={Classes.INTERACTIVE}
        elevation={Elevation.ONE}
        onClick={() =>
          handleContestEntryClick(contestEntry.submission_id, contestEntry.answer.code ?? '')
        }
      >
        <Pre className="contestentry-entryid">{entryNumber}</Pre>
        <Pre className="contestentry-rank">
          <NumericInput
            disabled={!canSave}
            value={contestEntry.score}
            onValueChange={(score: number) =>
              handleVotingSubmissionChange(contestEntry.submission_id, score)
            }
            placeholder={`Enter score for entry ${entryNumber}`}
            min={minScore}
            max={maxScore}
            allowNumericCharactersOnly
            fill
            minorStepSize={null} // limits input to integers
          />
        </Pre>
      </Card>
    </div>
  );
};

export default SideContentContestEntryCard;
