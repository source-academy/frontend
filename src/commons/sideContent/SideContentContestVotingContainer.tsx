import React, { useEffect, useState } from 'react';
import { showWarningMessage } from 'src/commons/utils/NotificationsHelper';

import { ContestEntry } from '../assessment/AssessmentTypes';
import SideContentContestVoting from './SideContentContestVoting';

export type SideContentContestVotingContainerProps = DispatchProps & StateProps;

type DispatchProps = {
  handleContestEntryClick: (submissionId: number, answer: string) => void;
  handleSave: (votingSubmission: ContestEntry[]) => void;
};

type StateProps = {
  canSave: boolean;
  contestEntries: ContestEntry[];
};

/**
 * Container to separate behaviour concerns from rendering concerns
 * Stores component-level voting ranking state
 */
const SideContentContestVotingContainer: React.FunctionComponent<
  SideContentContestVotingContainerProps
> = props => {
  const { canSave, contestEntries, handleSave, handleContestEntryClick } = props;
  const [isValid, setIsValid] = useState<boolean>(true);
  const [votingSubmission, setVotingSubmission] = useState<ContestEntry[]>([]);
  const minScore = 11 - contestEntries.length;
  const maxScore = 10;

  useEffect(() => {
    setVotingSubmission(contestEntries);
  }, [contestEntries]);

  /**
   * Validates input value and clamps the value within the min-max range [1, number of entries].
   * @param votingSubmission voting scores by user for each contest entry.
   * @returns boolean value for whether the scores are within the min-max range.
   */
  const isSubmissionValid = (votingSubmission: ContestEntry[]) => {
    return votingSubmission.reduce((isValid, vote) => {
      const score = vote.score!;
      return isValid && score >= minScore && score <= maxScore;
    }, true);
  };

  const submissionHasNoNull = (votingSubmission: ContestEntry[]) => {
    return votingSubmission.reduce((noNull, vote) => {
      return noNull && vote.score !== undefined && vote.score !== null;
    }, true);
  };

  const handleVotingSubmissionChange = (submissionId: number, score: number): void => {
    // update the votes
    const updatedSubmission = votingSubmission.map(vote =>
      vote.submission_id === submissionId ? { ...vote, score: score } : vote
    );
    setVotingSubmission(updatedSubmission);
    const noDuplicates =
      new Set(updatedSubmission.map(vote => vote.score)).size === updatedSubmission.length;
    // validate that scores are unique
    const noNull = submissionHasNoNull(updatedSubmission);
    if (noDuplicates && noNull && isSubmissionValid(updatedSubmission)) {
      handleSave(updatedSubmission);

      setIsValid(true);
    } else if (noDuplicates && noNull) {
      showWarningMessage(
        `Vote rankings invalid. Please input scores between ${minScore} - ${maxScore}.`
      );
      setIsValid(false);
    } else if (!noDuplicates && noNull) {
      showWarningMessage('Vote scores are not unique. Please input unique rankings.');
      setIsValid(false);
    } else {
      setIsValid(false);
    }
  };

  return (
    <SideContentContestVoting
      isValid={isValid}
      canSave={canSave}
      handleContestEntryClick={handleContestEntryClick}
      handleVotingSubmissionChange={handleVotingSubmissionChange}
      contestEntries={votingSubmission}
      minScore={minScore}
      maxScore={maxScore}
    />
  );
};

export default SideContentContestVotingContainer;
