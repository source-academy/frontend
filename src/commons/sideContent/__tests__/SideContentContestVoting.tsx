import { NumericInput } from '@blueprintjs/core';
import { Card } from '@blueprintjs/core';
import { mount } from 'enzyme';

import SideContentContestEntryCard from '../SideContentContestEntryCard';
import SideContentContestVotingContainer from '../SideContentContestVotingContainer';

const mockContestEntries = [
  {
    submission_id: 1,
    answer: { code: "display('hello world')" }
  },
  {
    submission_id: 2,
    answer: { code: 'function test() { return 1; }' }
  }
];

const mockProps = {
  handleContestEntryClick: () => {},
  handleSave: () => {},
  canSave: true,
  contestEntries: mockContestEntries
};

// Basic snapshot testing to catch unexpected changes
test('SideContentContestVotingContainer matches snapshot', () => {
  const contestVotingComponentRender = mount(<SideContentContestVotingContainer {...mockProps} />);

  expect(contestVotingComponentRender.debug()).toMatchSnapshot();
});

test('SideContentContestVotingContainer component renders correct number of entries.', () => {
  const contestVotingContainer = <SideContentContestVotingContainer {...mockProps} />;
  const contestVotingContainerRender = mount(contestVotingContainer);

  expect(contestVotingContainerRender.find('SideContentContestEntryCard')).toHaveLength(
    mockContestEntries.length
  );

  contestVotingContainerRender.setProps({
    contestEntries: [...mockContestEntries, { submission_id: 3, answer: { code: '' } }]
  });

  expect(contestVotingContainerRender.find('SideContentContestEntryCard')).toHaveLength(
    mockContestEntries.length
  );
});

test('SideContentVotingContainer disabled once canSave == false', () => {
  const mockProps = {
    handleContestEntryClick: () => {},
    handleSave: () => {},
    canSave: false,
    contestEntries: mockContestEntries
  };

  const contestVotingComponentRender = mount(<SideContentContestVotingContainer {...mockProps} />);
  const contestVotingCards = contestVotingComponentRender.find(NumericInput);
  contestVotingCards.map(card => expect(card.prop('disabled')).toBe(true));
});

// testing the ranking validation logic
test('SideContentVotingContainer only updates when ranks assigned to entries are unique and within rank limit.', () => {
  const mockedHandleContestEntryClick = jest.fn();
  const mockedHandleSave = jest.fn();

  const mockProps = {
    handleContestEntryClick: mockedHandleContestEntryClick,
    handleSave: mockedHandleSave,
    canSave: true,
    contestEntries: mockContestEntries
  };

  const contestVotingContainer = <SideContentContestVotingContainer {...mockProps} />;
  const contestVotingContainerRender = mount(contestVotingContainer);

  const contestVotingCard = contestVotingContainerRender.find('input');

  // simulate change to duplicate
  contestVotingCard.map(card => card.simulate('change', { target: { value: 10 } }));
  expect(mockedHandleSave).toHaveBeenCalledTimes(0);

  // to avoid warning messages from blueprint.js due to testing invalid input during tests
  const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});

  // simulate change to exceed score limit (ie. if 2 entries can only rank [9, 10])
  contestVotingCard.map(card => card.simulate('change', { target: { value: 11 } }));
  expect(mockedHandleSave).toHaveBeenCalledTimes(0);

  // simulate appropriate ranking for entries
  contestVotingCard.map((card, index) =>
    card.simulate('change', { target: { value: 10 - index } })
  );
  expect(mockedHandleSave).toHaveBeenCalledTimes(1);

  consoleWarnSpy.mockRestore();
});

test('Clicking the contest entry updates the editor for contest voting.', () => {
  const mockedHandleContestEntryClick = jest.fn();

  const mockProps = {
    handleContestEntryClick: mockedHandleContestEntryClick,
    handleSave: () => {},
    canSave: true,
    contestEntries: mockContestEntries
  };

  const contestVotingContainer = <SideContentContestVotingContainer {...mockProps} />;
  const contestVotingContainerRender = mount(contestVotingContainer);

  mockedHandleContestEntryClick.mockClear();
  contestVotingContainerRender.find(SideContentContestEntryCard).find(Card).at(0).simulate('click');
  expect(mockedHandleContestEntryClick).toBeCalledTimes(1);
  expect(mockedHandleContestEntryClick.mock.calls[0][0]).toBe(1);
  expect(mockedHandleContestEntryClick.mock.calls[0][1]).toBe("display('hello world')");

  contestVotingContainerRender.find(SideContentContestEntryCard).find(Card).at(1).simulate('click');
  expect(mockedHandleContestEntryClick).toBeCalledTimes(2);
  expect(mockedHandleContestEntryClick.mock.calls[1][0]).toBe(2);
  expect(mockedHandleContestEntryClick.mock.calls[1][1]).toBe('function test() { return 1; }');
});
