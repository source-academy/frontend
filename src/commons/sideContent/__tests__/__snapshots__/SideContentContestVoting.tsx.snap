// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideContentContestVotingContainer matches snapshot 1`] = `
"<SideContentContestVotingContainer handleContestEntryClick={[Function: handleContestEntryClick]} handleSave={[Function: handleSave]} canSave={true} contestEntries={{...}}>
  <SideContentContestVoting isValid={true} canSave={true} handleContestEntryClick={[Function: handleContestEntryClick]} handleVotingSubmissionChange={[Function: handleVotingSubmissionChange]} contestEntries={{...}} minScore={9} maxScore={10}>
    <div className=\\"ContestEntryVoting\\">
      <Blueprint4.Button className=\\"collapse-button\\" icon=\\"caret-down\\" minimal={true} onClick={[Function: onClick]}>
        <button type=\\"button\\" className=\\"bp4-button bp4-minimal collapse-button\\" onClick={[Function: onClick]} disabled={false} onBlur={[Function (anonymous)]} onFocus={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
          <Blueprint4.Icon icon=\\"caret-down\\">
            <span aria-hidden={true} icon=\\"caret-down\\" className=\\"bp4-icon bp4-icon-caret-down\\" title={[undefined]}>
              <svg fill={[undefined]} data-icon=\\"caret-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                <path d=\\"M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z\\" fillRule=\\"evenodd\\" />
              </svg>
            </span>
          </Blueprint4.Icon>
          <span className=\\"bp4-button-text\\">
            <span>
              Contest Voting
            </span>
            <Blueprint4.Tooltip2 content={{...}} compact={false} hoverCloseDelay={0} hoverOpenDelay={100} interactionKind=\\"hover-target\\" minimal={false} transitionDuration={100}>
              <Tooltip2Provider>
                <Blueprint4.Popover2 modifiers={{...}} content={{...}} hoverCloseDelay={0} hoverOpenDelay={100} interactionKind=\\"hover-target\\" minimal={false} transitionDuration={100} autoFocus={false} canEscapeKeyClose={false} disabled={false} enforceFocus={false} lazy={true} popoverClassName=\\"bp4-tooltip2\\" portalContainer={[undefined]} boundary=\\"clippingParents\\" captureDismiss={false} defaultIsOpen={false} fill={false} hasBackdrop={false} inheritDarkTheme={true} matchTargetWidth={false} openOnTargetFocus={true} positioningStrategy=\\"absolute\\" renderTarget={[undefined]} shouldReturnFocusOnClose={false} targetTagName=\\"span\\" usePortal={true}>
                  <Manager>
                    <Reference>
                      <Blueprint4.ResizeSensor2 targetRef={[Function (anonymous)]} onResize={[Function (anonymous)]}>
                        <span aria-haspopup={[undefined]} className=\\"bp4-popover2-target\\" onBlur={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onFocus={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]}>
                          <Blueprint4.Icon icon=\\"help\\" className=\\"\\" disabled={[undefined]} tabIndex={0}>
                            <span aria-hidden={true} icon=\\"help\\" disabled={[undefined]} tabIndex={0} className=\\"bp4-icon bp4-icon-help\\" title={[undefined]}>
                              <svg fill={[undefined]} data-icon=\\"help\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                <path d=\\"M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13H7v-2h2v2zm1.93-6.52c-.14.32-.35.64-.62.97L9.25 8.83c-.12.15-.24.29-.28.42-.04.13-.09.3-.09.52V10H7.12V8.88s.05-.51.21-.71L8.4 6.73c.22-.26.35-.49.44-.68.09-.19.12-.38.12-.58 0-.3-.1-.55-.28-.75-.18-.19-.44-.28-.76-.28-.33 0-.59.1-.78.29-.19.19-.33.46-.4.81-.03.11-.1.15-.2.14l-1.7-.25c-.12-.01-.16-.08-.14-.19.12-.82.46-1.47 1.03-1.94.57-.48 1.32-.72 2.25-.72.47 0 .9.07 1.29.22s.72.34 1 .59c.28.25.49.55.65.89.15.35.22.72.22 1.12s-.07.75-.21 1.08z\\" fillRule=\\"evenodd\\" />
                              </svg>
                            </span>
                          </Blueprint4.Icon>
                        </span>
                      </Blueprint4.ResizeSensor2>
                    </Reference>
                    <Popper innerRef={[Function (anonymous)]} placement=\\"auto\\" strategy=\\"absolute\\" modifiers={{...}}>
                      <Blueprint4.Overlay autoFocus={false} backdropClassName=\\"bp4-popover2-backdrop\\" backdropProps={{...}} canEscapeKeyClose={false} canOutsideClickClose={false} enforceFocus={false} hasBackdrop={false} isOpen={false} onClose={[Function (anonymous)]} onClosed={[undefined]} onClosing={[undefined]} onOpened={[undefined]} onOpening={[undefined]} transitionDuration={100} transitionName=\\"bp4-popover2\\" usePortal={true} portalClassName={[undefined]} portalContainer={[undefined]} shouldReturnFocusOnClose={false} lazy={true} />
                    </Popper>
                  </Manager>
                </Blueprint4.Popover2>
              </Tooltip2Provider>
            </Blueprint4.Tooltip2>
          </span>
          <Blueprint4.Icon icon={[undefined]} />
        </button>
      </Blueprint4.Button>
      <Blueprint4.Collapse isOpen={true} keepChildrenMounted={true} component=\\"div\\" transitionDuration={200}>
        <div className=\\"bp4-collapse\\" style={{...}}>
          <div className=\\"bp4-collapse-body\\" style={{...}} aria-hidden={false}>
            <div>
              <div className=\\"contestentries-header\\">
                <div className=\\"header-entryid\\">
                  Entry Id
                  <Blueprint4.Icon icon=\\"caret-down\\">
                    <span aria-hidden={true} icon=\\"caret-down\\" className=\\"bp4-icon bp4-icon-caret-down\\" title={[undefined]}>
                      <svg fill={[undefined]} data-icon=\\"caret-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                        <path d=\\"M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z\\" fillRule=\\"evenodd\\" />
                      </svg>
                    </span>
                  </Blueprint4.Icon>
                </div>
                <div className=\\"header-entryrank\\">
                  Score
                  <Blueprint4.Icon icon=\\"caret-down\\">
                    <span aria-hidden={true} icon=\\"caret-down\\" className=\\"bp4-icon bp4-icon-caret-down\\" title={[undefined]}>
                      <svg fill={[undefined]} data-icon=\\"caret-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                        <path d=\\"M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z\\" fillRule=\\"evenodd\\" />
                      </svg>
                    </span>
                  </Blueprint4.Icon>
                </div>
              </div>
              <SideContentContestEntryCard isValid={true} canSave={true} entryNumber={1} handleContestEntryClick={[Function: handleContestEntryClick]} handleVotingSubmissionChange={[Function: handleVotingSubmissionChange]} contestEntry={{...}} minScore={9} maxScore={10}>
                <div className=\\"ContestEntryCard\\">
                  <Blueprint4.Card className=\\"bp4-interactive\\" elevation={1} onClick={[Function: onClick]} interactive={false}>
                    <div className=\\"bp4-card bp4-elevation-1 bp4-interactive\\" onClick={[Function: onClick]}>
                      <Component className=\\"contestentry-entryid\\">
                        <pre className=\\"bp4-code-block contestentry-entryid\\">
                          1
                        </pre>
                      </Component>
                      <Component className=\\"contestentry-rank\\">
                        <pre className=\\"bp4-code-block contestentry-rank\\">
                          <Blueprint4.NumericInput disabled={false} value={[undefined]} onValueChange={[Function: onValueChange]} placeholder=\\"Enter score for entry 1\\" min={9} max={10} allowNumericCharactersOnly={true} fill={true} minorStepSize={{...}} buttonPosition=\\"right\\" clampValueOnBlur={false} defaultValue=\\"\\" large={false} majorStepSize={10} selectAllOnFocus={false} selectAllOnIncrement={false} stepSize={1}>
                            <Blueprint4.ControlGroup className=\\"bp4-numeric-input\\" fill={true}>
                              <div className=\\"bp4-control-group bp4-fill bp4-numeric-input\\">
                                <Blueprint4.InputGroup asyncControl={[undefined]} autoComplete=\\"off\\" id=\\"numericInput-0\\" role=\\"spinbutton\\" disabled={false} value=\\"\\" placeholder=\\"Enter score for entry 1\\" min={9} max={10} aria-valuemax={10} aria-valuemin={9} aria-valuenow={0} intent={[undefined]} inputClassName={[undefined]} inputRef={[Function (anonymous)]} large={false} leftElement={[undefined]} leftIcon={[undefined]} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onCompositionEnd={[Function (anonymous)]} onCompositionUpdate={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyPress={[Function (anonymous)]} onPaste={[Function (anonymous)]} rightElement={[undefined]}>
                                  <div className=\\"bp4-input-group\\">
                                    <input type=\\"text\\" autoComplete=\\"off\\" id=\\"numericInput-0\\" role=\\"spinbutton\\" disabled={false} value=\\"\\" placeholder=\\"Enter score for entry 1\\" min={9} max={10} aria-valuemax={10} aria-valuemin={9} aria-valuenow={0} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onCompositionEnd={[Function (anonymous)]} onCompositionUpdate={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyPress={[Function (anonymous)]} onPaste={[Function (anonymous)]} className=\\"bp4-input\\" style={{...}} />
                                  </div>
                                </Blueprint4.InputGroup>
                                <Blueprint4.ButtonGroup className=\\"bp4-fixed\\" vertical={true}>
                                  <div className=\\"bp4-button-group bp4-vertical bp4-fixed\\">
                                    <Blueprint4.Button aria-label=\\"increment\\" aria-controls=\\"numericInput-0\\" disabled={false} icon=\\"chevron-up\\" intent={[undefined]} onKeyDown={[Function: onKeyDown]} onMouseDown={[Function: onMouseDown]}>
                                      <button type=\\"button\\" aria-label=\\"increment\\" aria-controls=\\"numericInput-0\\" disabled={false} onKeyDown={[Function (anonymous)]} onMouseDown={[Function: onMouseDown]} className=\\"bp4-button\\" onBlur={[Function (anonymous)]} onClick={[undefined]} onFocus={[undefined]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                        <Blueprint4.Icon icon=\\"chevron-up\\">
                                          <span aria-hidden={true} icon=\\"chevron-up\\" className=\\"bp4-icon bp4-icon-chevron-up\\" title={[undefined]}>
                                            <svg fill={[undefined]} data-icon=\\"chevron-up\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                              <path d=\\"M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 001.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71z\\" fillRule=\\"evenodd\\" />
                                            </svg>
                                          </span>
                                        </Blueprint4.Icon>
                                        <Blueprint4.Icon icon={[undefined]} />
                                      </button>
                                    </Blueprint4.Button>
                                    <Blueprint4.Button aria-label=\\"decrement\\" aria-controls=\\"numericInput-0\\" disabled={false} icon=\\"chevron-down\\" intent={[undefined]} onKeyDown={[Function: onKeyDown]} onMouseDown={[Function: onMouseDown]}>
                                      <button type=\\"button\\" aria-label=\\"decrement\\" aria-controls=\\"numericInput-0\\" disabled={false} onKeyDown={[Function (anonymous)]} onMouseDown={[Function: onMouseDown]} className=\\"bp4-button\\" onBlur={[Function (anonymous)]} onClick={[undefined]} onFocus={[undefined]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                        <Blueprint4.Icon icon=\\"chevron-down\\">
                                          <span aria-hidden={true} icon=\\"chevron-down\\" className=\\"bp4-icon bp4-icon-chevron-down\\" title={[undefined]}>
                                            <svg fill={[undefined]} data-icon=\\"chevron-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                              <path d=\\"M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z\\" fillRule=\\"evenodd\\" />
                                            </svg>
                                          </span>
                                        </Blueprint4.Icon>
                                        <Blueprint4.Icon icon={[undefined]} />
                                      </button>
                                    </Blueprint4.Button>
                                  </div>
                                </Blueprint4.ButtonGroup>
                              </div>
                            </Blueprint4.ControlGroup>
                          </Blueprint4.NumericInput>
                        </pre>
                      </Component>
                    </div>
                  </Blueprint4.Card>
                </div>
              </SideContentContestEntryCard>
              <SideContentContestEntryCard isValid={true} canSave={true} entryNumber={2} handleContestEntryClick={[Function: handleContestEntryClick]} handleVotingSubmissionChange={[Function: handleVotingSubmissionChange]} contestEntry={{...}} minScore={9} maxScore={10}>
                <div className=\\"ContestEntryCard\\">
                  <Blueprint4.Card className=\\"bp4-interactive\\" elevation={1} onClick={[Function: onClick]} interactive={false}>
                    <div className=\\"bp4-card bp4-elevation-1 bp4-interactive\\" onClick={[Function: onClick]}>
                      <Component className=\\"contestentry-entryid\\">
                        <pre className=\\"bp4-code-block contestentry-entryid\\">
                          2
                        </pre>
                      </Component>
                      <Component className=\\"contestentry-rank\\">
                        <pre className=\\"bp4-code-block contestentry-rank\\">
                          <Blueprint4.NumericInput disabled={false} value={[undefined]} onValueChange={[Function: onValueChange]} placeholder=\\"Enter score for entry 2\\" min={9} max={10} allowNumericCharactersOnly={true} fill={true} minorStepSize={{...}} buttonPosition=\\"right\\" clampValueOnBlur={false} defaultValue=\\"\\" large={false} majorStepSize={10} selectAllOnFocus={false} selectAllOnIncrement={false} stepSize={1}>
                            <Blueprint4.ControlGroup className=\\"bp4-numeric-input\\" fill={true}>
                              <div className=\\"bp4-control-group bp4-fill bp4-numeric-input\\">
                                <Blueprint4.InputGroup asyncControl={[undefined]} autoComplete=\\"off\\" id=\\"numericInput-1\\" role=\\"spinbutton\\" disabled={false} value=\\"\\" placeholder=\\"Enter score for entry 2\\" min={9} max={10} aria-valuemax={10} aria-valuemin={9} aria-valuenow={0} intent={[undefined]} inputClassName={[undefined]} inputRef={[Function (anonymous)]} large={false} leftElement={[undefined]} leftIcon={[undefined]} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onCompositionEnd={[Function (anonymous)]} onCompositionUpdate={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyPress={[Function (anonymous)]} onPaste={[Function (anonymous)]} rightElement={[undefined]}>
                                  <div className=\\"bp4-input-group\\">
                                    <input type=\\"text\\" autoComplete=\\"off\\" id=\\"numericInput-1\\" role=\\"spinbutton\\" disabled={false} value=\\"\\" placeholder=\\"Enter score for entry 2\\" min={9} max={10} aria-valuemax={10} aria-valuemin={9} aria-valuenow={0} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onCompositionEnd={[Function (anonymous)]} onCompositionUpdate={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyPress={[Function (anonymous)]} onPaste={[Function (anonymous)]} className=\\"bp4-input\\" style={{...}} />
                                  </div>
                                </Blueprint4.InputGroup>
                                <Blueprint4.ButtonGroup className=\\"bp4-fixed\\" vertical={true}>
                                  <div className=\\"bp4-button-group bp4-vertical bp4-fixed\\">
                                    <Blueprint4.Button aria-label=\\"increment\\" aria-controls=\\"numericInput-1\\" disabled={false} icon=\\"chevron-up\\" intent={[undefined]} onKeyDown={[Function: onKeyDown]} onMouseDown={[Function: onMouseDown]}>
                                      <button type=\\"button\\" aria-label=\\"increment\\" aria-controls=\\"numericInput-1\\" disabled={false} onKeyDown={[Function (anonymous)]} onMouseDown={[Function: onMouseDown]} className=\\"bp4-button\\" onBlur={[Function (anonymous)]} onClick={[undefined]} onFocus={[undefined]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                        <Blueprint4.Icon icon=\\"chevron-up\\">
                                          <span aria-hidden={true} icon=\\"chevron-up\\" className=\\"bp4-icon bp4-icon-chevron-up\\" title={[undefined]}>
                                            <svg fill={[undefined]} data-icon=\\"chevron-up\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                              <path d=\\"M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 001.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71z\\" fillRule=\\"evenodd\\" />
                                            </svg>
                                          </span>
                                        </Blueprint4.Icon>
                                        <Blueprint4.Icon icon={[undefined]} />
                                      </button>
                                    </Blueprint4.Button>
                                    <Blueprint4.Button aria-label=\\"decrement\\" aria-controls=\\"numericInput-1\\" disabled={false} icon=\\"chevron-down\\" intent={[undefined]} onKeyDown={[Function: onKeyDown]} onMouseDown={[Function: onMouseDown]}>
                                      <button type=\\"button\\" aria-label=\\"decrement\\" aria-controls=\\"numericInput-1\\" disabled={false} onKeyDown={[Function (anonymous)]} onMouseDown={[Function: onMouseDown]} className=\\"bp4-button\\" onBlur={[Function (anonymous)]} onClick={[undefined]} onFocus={[undefined]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                        <Blueprint4.Icon icon=\\"chevron-down\\">
                                          <span aria-hidden={true} icon=\\"chevron-down\\" className=\\"bp4-icon bp4-icon-chevron-down\\" title={[undefined]}>
                                            <svg fill={[undefined]} data-icon=\\"chevron-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                              <path d=\\"M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z\\" fillRule=\\"evenodd\\" />
                                            </svg>
                                          </span>
                                        </Blueprint4.Icon>
                                        <Blueprint4.Icon icon={[undefined]} />
                                      </button>
                                    </Blueprint4.Button>
                                  </div>
                                </Blueprint4.ButtonGroup>
                              </div>
                            </Blueprint4.ControlGroup>
                          </Blueprint4.NumericInput>
                        </pre>
                      </Component>
                    </div>
                  </Blueprint4.Card>
                </div>
              </SideContentContestEntryCard>
            </div>
          </div>
        </div>
      </Blueprint4.Collapse>
    </div>
  </SideContentContestVoting>
</SideContentContestVotingContainer>"
`;
